cmd/update/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! ## `humility update`
+//!
+//! Writes a binary to the specified update target as defined in Hubris
+//!
+//! ```console
+//! $ humility update --target ImageB update.bin
+//! humility: attached via CMSIS-DAP
+//! humility: Starting update using an update block size of 512
+//! humility: (Erase may take a moment)
+//! humility: Comitting update
+//! humility: Update done.
+//! ```
+//!
+
+use humility::cli::Subcommand;
+use humility_cmd::hiffy::*;
+use humility_cmd::{Archive, Attach, Command, Validate};
+
+use std::path::PathBuf;
+
+use anyhow::{bail, Result};
+use clap::Command as ClapCommand;
+use clap::{CommandFactory, Parser};
+use hif::*;
+
+use indicatif::{ProgressBar, ProgressStyle};
+
+extern crate log;
+
+#[derive(Parser, Debug)]
+#[clap(name = "update", about = "Write a software update")]
+struct UpdateArgs {
+    /// sets timeout
+    #[clap(
+        long, short = 'T', default_value = "50000", value_name = "timeout_ms",
+        parse(try_from_str = parse_int::parse)
+    )]
+    timeout: u32,
+
+    #[clap(long)]
+    /// Which image to update
+    target: String,
+    #[clap(long)]
+    /// Don't commit the image, only write
+    skip_commit: bool,
+    /// path to binary to write
+    path: PathBuf,
+}
+
+fn update(context: &mut humility::ExecutionContext) -> Result<()> {
+    let core = &mut **context.core.as_mut().unwrap();
+    let hubris = context.archive.as_ref().unwrap();
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
+
+    let subargs = UpdateArgs::try_parse_from(subargs)?;
+    let mut context = HiffyContext::new(hubris, core, subargs.timeout)?;
+    let funcs = context.functions()?;
+
+    let start_update_fn = funcs.get("StartUpdate", 0)?;
+    let write_block_fn = funcs.get("WriteBlock", 2)?;
+    let finish_update_fn = funcs.get("FinishUpdate", 0)?;
+    let block_size_fn = funcs.get("BlockSize", 0)?;
+
+    let t = hubris.lookup_task("update_server").unwrap();
+    let m = hubris.lookup_module(*t)?;
+    let update_enum = m.lookup_enum_byname(hubris, "UpdateTarget")?;
+
+    let update_target = match update_enum.lookup_variant_byname(&subargs.target)
+    {
+        Ok(v) => v.tag.unwrap(),
+        _ => bail!("Couldn't find image target {}", subargs.target),
+    };
+
+    let ops = vec![Op::Call(block_size_fn.id), Op::Done];
+
+    let results = context.run(core, ops.as_slice(), None)?;
+
+    let blk_size = match &results[0] {
+        Ok(s) => {
+            if s.len() < 4 {
+                bail!("Bad length for block size!");
+            }
+            (s[0] as u32)
```

Comment:
If we know that `s` is exactly 4 `u8`s long, then
```rust
u32::from_le_bytes(s.try_into().unwrap())
```
otherwise
```rust
u32::from_le_bytes(s[..4].try_into().unwrap())
```